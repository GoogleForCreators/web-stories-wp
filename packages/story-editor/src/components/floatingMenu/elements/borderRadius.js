/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { Icons } from '@googleforcreators/design-system';
import { useRef } from '@googleforcreators/react';
import { __ } from '@googleforcreators/i18n';
import { canSupportMultiBorder } from '@googleforcreators/masks';
import { trackEvent } from '@googleforcreators/tracking';

/**
 * Internal dependencies
 */
import { useStory } from '../../../app';
import { DEFAULT_BORDER_RADIUS } from '../../panels/design/sizePosition/radius';
import {
  Input,
  Separator,
  useProperties,
  FocusTrapButton,
  handleReturnTrappedFocus,
} from './shared';

const CORNER_LABEL = __('Corner Radius', 'web-stories');
function BorderRadius() {
  const inputRef = useRef();
  const buttonRef = useRef();
  // Note that "mask" never updates on an element,
  // so selecting it cannot cause re-renders
  // We need it to determine if radii are supported.
  const {
    borderRadius = DEFAULT_BORDER_RADIUS,
    mask,
    type,
  } = useProperties(['borderRadius', 'mask', 'type']);

  const { selectedElementIds, updateElementsById } = useStory(
    ({ state, actions }) => ({
      selectedElementIds: state.selectedElementIds,
      updateElementsById: actions.updateElementsById,
    })
  );

  // Only multi-border elements support border radius
  const canHaveBorderRadius = canSupportMultiBorder({ mask });

  // We only allow editing the current border radii, if all corners are identical
  const hasUniformBorder =
    borderRadius.topLeft === borderRadius.topRight &&
    borderRadius.topLeft === borderRadius.bottomLeft &&
    borderRadius.topLeft === borderRadius.bottomRight;

  // Render nothing if radii not supported or not locked
  if (!canHaveBorderRadius || (!borderRadius.locked && !hasUniformBorder)) {
    return null;
  }

  const onChange = (value) => {
    void trackEvent('floating_menu', {
      name: 'set_border_radius',
      element: type,
    });

    updateElementsById({
      elementIds: selectedElementIds,
      properties: {
        borderRadius: {
          locked: true,
          topLeft: value,
          topRight: value,
          bottomRight: value,
          bottomLeft: value,
        },
      },
    });
  };

  return (
    <FocusTrapButton
      ref={buttonRef}
      inputRef={inputRef}
      inputLabel={CORNER_LABEL}
    >
      <Input
        tabIndex={-1}
        ref={inputRef}
        suffix={<Icons.Corner />}
        value={borderRadius.topLeft}
        aria-label={CORNER_LABEL}
        onChange={(_, value) => onChange(value)}
        onKeyDown={(e) => {
          handleReturnTrappedFocus(e, buttonRef);
        }}
      />
      <Separator />
    </FocusTrapButton>
  );
}

export default BorderRadius;
