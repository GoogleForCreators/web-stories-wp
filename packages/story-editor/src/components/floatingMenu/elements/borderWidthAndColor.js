/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { Icons } from '@googleforcreators/design-system';
import { useRef } from '@googleforcreators/react';
import { __ } from '@googleforcreators/i18n';
import { trackEvent } from '@googleforcreators/tracking';
import styled from 'styled-components';
import { canSupportMultiBorder } from '@googleforcreators/masks';

/**
 * Internal dependencies
 */
import { useStory } from '../../../app';
import { DEFAULT_BORDER } from '../../panels/design/border/shared';
import {
  Input,
  Color,
  useProperties,
  FocusTrapButton,
  handleReturnTrappedFocus,
} from './shared';

const Container = styled.div`
  display: flex;
  align-items: center;
`;

const Dash = styled.div`
  width: 8px;
  height: 1px;
  background-color: ${({ theme }) => theme.colors.border.defaultNormal};
`;

function getHasBorderWidth(border) {
  if (!border) {
    return false;
  }
  const { left, right, top, bottom } = border;
  return Boolean(left || right || top || bottom);
}

const BLACK = { color: { r: 0, g: 0, b: 0 } };

const WIDTH_LABEL = __('Border width', 'web-stories');
const COLOR_LABEL = __('Border color', 'web-stories');

function BorderWidthAndColor() {
  const inputRef = useRef();
  const buttonRef = useRef();
  // Note that "mask" never updates on an element,
  // so selecting it cannot cause re-renders
  // We need it to determine if border opacity is allowed.
  const {
    border = DEFAULT_BORDER,
    mask,
    type,
  } = useProperties(['border', 'mask', 'type']);

  const { selectedElementIds, updateElementsById } = useStory(
    ({ state, actions }) => ({
      selectedElementIds: state.selectedElementIds,
      updateElementsById: actions.updateElementsById,
    })
  );

  // We only allow editing the current border width, if all borders are identical
  const hasUniformBorder =
    border?.left === border?.right &&
    border?.left === border?.top &&
    border?.left === border?.bottom;

  // Border width and color inputs should only be rendered if all sides of the border are the same.
  // Both checks are needed since not all shaped have lockedWidth eg. circles.
  if (!border?.lockedWidth && !hasUniformBorder) {
    return null;
  }

  // Only multi-border elements support border opacity
  const canHaveBorderOpacity = canSupportMultiBorder({ mask });

  // We only allow editing border color, if at least one border has a non-zero width
  const hasBorderWidth = getHasBorderWidth(border);

  const handleWidthChange = (value) => {
    void trackEvent('floating_menu', {
      name: 'set_border_width',
      element: type,
    });

    updateElementsById({
      elementIds: selectedElementIds,
      properties: ({ border: oldBorder }) => ({
        border: {
          locked: true,
          color: border?.color,
          ...oldBorder,
          left: value,
          right: value,
          top: value,
          bottom: value,
        },
      }),
    });
  };

  const handleColorChange = (value) => {
    void trackEvent('floating_menu', {
      name: 'set_border_color',
      element: type,
    });

    updateElementsById({
      elementIds: selectedElementIds,
      properties: ({ border: oldBorder }) => ({
        border: {
          ...oldBorder,
          color: value,
        },
      }),
    });
  };

  return (
    <Container>
      <FocusTrapButton
        ref={buttonRef}
        inputRef={inputRef}
        inputLabel={WIDTH_LABEL}
      >
        <Input
          tabIndex={-1}
          ref={inputRef}
          suffix={<Icons.BorderBox />}
          value={border?.left || 0}
          aria-label={WIDTH_LABEL}
          onChange={(_, value) => handleWidthChange(value)}
          onKeyDown={(e) => {
            handleReturnTrappedFocus(e, buttonRef);
          }}
        />
      </FocusTrapButton>
      {hasBorderWidth && (
        <>
          <Dash />
          <Color
            tabIndex={-1}
            label={COLOR_LABEL}
            value={border?.color || BLACK}
            onChange={handleColorChange}
            hasInputs={false}
            hasEyedropper={false}
            allowsOpacity={canHaveBorderOpacity}
            allowsGradient={false}
            allowsSavedColors={false}
          />
        </>
      )}
    </Container>
  );
}

export default BorderWidthAndColor;
