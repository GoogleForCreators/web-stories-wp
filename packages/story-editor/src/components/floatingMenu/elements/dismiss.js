/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { Icons } from '@googleforcreators/design-system';
import { __ } from '@googleforcreators/i18n';
import { trackEvent } from '@googleforcreators/tracking';
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useFloatingMenu } from '../context';
import { IconButton, useProperties } from './shared';

function Dismiss({ isMultiple }) {
  const { type } = useProperties(['type']);
  const { handleDismiss } = useFloatingMenu(
    ({ actions: { handleDismiss } }) => ({
      handleDismiss,
    })
  );

  const handleClick = () => {
    trackEvent('floating_menu', {
      name: 'dismiss_menu',
      element: isMultiple ? 'multiple' : type,
    });
    handleDismiss();
  };

  return (
    <IconButton
      Icon={Icons.Cross}
      title={__('Dismiss menu', 'web-stories')}
      onClick={handleClick}
    />
  );
}

export default Dismiss;

Dismiss.propTypes = {
  isMultiple: PropTypes.bool,
};
