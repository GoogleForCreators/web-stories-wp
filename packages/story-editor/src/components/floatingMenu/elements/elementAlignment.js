/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { Icons } from '@googleforcreators/design-system';
import { __ } from '@googleforcreators/i18n';

/**
 * Internal dependencies
 */
import { useStory } from '../../../app';
import useAlignment from '../../panels/design/alignment/useAlignment';
import { IconButton, Separator } from './shared';

function ElementAlignment() {
  const { selectedElements, selectedElementIds, updateElementsById } = useStory(
    ({ state, actions }) => ({
      selectedElements: state.selectedElements,
      selectedElementIds: state.selectedElementIds,
      updateElementsById: actions.updateElementsById,
    })
  );

  const updateElements = (callback) => {
    updateElementsById({
      elementIds: selectedElementIds,
      properties: callback,
    });
  };

  const {
    isDistributionEnabled,
    handleAlignLeft,
    handleAlignCenter,
    handleAlignRight,
    handleAlignTop,
    handleAlignMiddle,
    handleAlignBottom,
    handleHorizontalDistribution,
    handleVerticalDistribution,
  } = useAlignment({ selectedElements, updateElements, isFloatingMenu: true });
  return (
    <>
      {isDistributionEnabled && (
        <>
          <IconButton
            Icon={Icons.DistributeHorizontal}
            title={__('Distribute horizontally', 'web-stories')}
            onClick={handleHorizontalDistribution}
          />
          <IconButton
            Icon={Icons.DistributeVertical}
            title={__('Distribute vertically', 'web-stories')}
            onClick={handleVerticalDistribution}
          />
          <Separator />
        </>
      )}
      <IconButton
        Icon={Icons.AlignLeft}
        title={__('Align left', 'web-stories')}
        onClick={handleAlignLeft}
      />
      <IconButton
        Icon={Icons.AlignCenter}
        title={__('Align center', 'web-stories')}
        onClick={handleAlignCenter}
      />
      <IconButton
        Icon={Icons.AlignRight}
        title={__('Align right', 'web-stories')}
        onClick={handleAlignRight}
      />
      <IconButton
        Icon={Icons.AlignTop}
        title={__('Align top', 'web-stories')}
        onClick={handleAlignTop}
      />
      <IconButton
        Icon={Icons.AlignMiddle}
        title={__('Align vertical center', 'web-stories')}
        onClick={handleAlignMiddle}
      />
      <IconButton
        Icon={Icons.AlignBottom}
        title={__('Align bottom', 'web-stories')}
        onClick={handleAlignBottom}
      />
    </>
  );
}

export default ElementAlignment;
