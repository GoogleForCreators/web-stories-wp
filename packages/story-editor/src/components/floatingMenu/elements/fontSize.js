/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { __ } from '@googleforcreators/i18n';
import styled from 'styled-components';
import { useCallback, useRef } from '@googleforcreators/react';
import { trackEvent } from '@googleforcreators/tracking';
import {
  NumericInput,
  CONTEXT_MENU_SKIP_ELEMENT,
} from '@googleforcreators/design-system';

/**
 * Internal dependencies
 */
import { useStory } from '../../../app';
import getUpdatedSizeAndPosition from '../../../utils/getUpdatedSizeAndPosition';
import updateProperties from '../../style/updateProperties';
import {
  focusStyle,
  inputContainerStyleOverride,
} from '../../panels/shared/styles';
import { MIN_MAX } from '../../panels/design/textStyle/font';
import { FocusTrapButton, handleReturnTrappedFocus } from './shared';

const Input = styled(NumericInput).attrs({
  inputClassName: CONTEXT_MENU_SKIP_ELEMENT,
})`
  width: 50px;
  flex: 0 0 50px;
`;

const FONT_SIZE_LABEL = __('Font size', 'web-stories');

function FontSize() {
  const inputRef = useRef();
  const buttonRef = useRef();
  const { fontSize, selectedElementIds, updateElementsById } = useStory(
    ({ state, actions }) => ({
      fontSize: state.selectedElements[0].fontSize,
      selectedElementIds: state.selectedElementIds,
      updateElementsById: actions.updateElementsById,
    })
  );

  const onChange = useCallback(
    (update) => {
      void trackEvent('floating_menu', {
        name: 'set_font_size',
      });

      updateElementsById({
        elementIds: selectedElementIds,
        properties: (element) => {
          const updates = updateProperties(element, update, true);
          const sizeUpdates = getUpdatedSizeAndPosition({
            ...element,
            ...updates,
          });
          return {
            ...updates,
            ...sizeUpdates,
          };
        },
      });
    },
    [selectedElementIds, updateElementsById]
  );

  return (
    <FocusTrapButton
      ref={buttonRef}
      inputRef={inputRef}
      inputLabel={FONT_SIZE_LABEL}
    >
      <Input
        tabIndex={-1}
        ref={inputRef}
        aria-label={FONT_SIZE_LABEL}
        isFloat
        value={fontSize}
        onChange={(evt, value) => onChange({ fontSize: value })}
        min={MIN_MAX.FONT_SIZE.MIN}
        max={MIN_MAX.FONT_SIZE.MAX}
        placeholder={fontSize}
        containerStyleOverride={inputContainerStyleOverride}
        selectButtonStylesOverride={focusStyle}
        onKeyDown={(e) => {
          handleReturnTrappedFocus(e, buttonRef);
        }}
      />
    </FocusTrapButton>
  );
}

export default FontSize;
