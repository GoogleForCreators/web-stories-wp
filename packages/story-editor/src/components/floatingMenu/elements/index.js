/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as BorderWidthAndColor } from './borderWidthAndColor';
export { default as BorderRadius } from './borderRadius';
export { default as Dismiss } from './dismiss';
export { default as ElementAlignment } from './elementAlignment';
export { default as FlipHorizontal } from './flipHorizontal';
export { default as FlipVertical } from './flipVertical';
export { default as FontFamily } from './fontFamily';
export { default as FontSize } from './fontSize';
export { default as LayerOpacity } from './layerOpacity';
export { default as Loop } from './loop';
export { default as More } from './more';
export { default as Mute } from './mute';
export { default as ShapeColor } from './shapeColor';
export { default as Settings } from './settings';
export { default as TextAlign } from './textAlign';
export { default as TextColor } from './textColor';
export { default as ToggleBold } from './toggleBold';
export { default as ToggleItalics } from './toggleItalics';
export { default as ToggleUnderline } from './toggleUnderline';
export { default as Trim } from './trim';
export { Separator } from './shared';
export { default as Trash } from './trash';
