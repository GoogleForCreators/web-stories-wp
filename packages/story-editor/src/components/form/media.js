/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { useCallback, forwardRef } from '@googleforcreators/react';
import { __ } from '@googleforcreators/i18n';
import styled from 'styled-components';
import {
  MediaInput as Input,
  themeHelpers,
} from '@googleforcreators/design-system';
/**
 * Internal dependencies
 */
import { useConfig } from '../../app';
import { MULTIPLE_VALUE } from '../../constants';
import HotlinkModal from '../hotlinkModal';
import useHotlink from './useHotlink';

const StyledInput = styled(Input)`
  button:focus {
    ${({ theme }) =>
      themeHelpers.focusCSS(
        theme.colors.border.focus,
        theme.colors.bg.secondary
      )}
  }
`;

const MediaInputField = ({
  open,
  openHotlink,
  dropdownOptions,
  onChange,
  forwardedRef,
  value,
  ...rest
}) => {
  const onMenuOption = useCallback(
    (evt, opt) => {
      switch (opt) {
        case 'upload':
          open(evt);
          break;
        case 'hotlink':
          openHotlink(evt);
          break;
        case 'remove':
        case 'reset':
          onChange(null);
          break;
        default:
          break;
      }
    },
    [onChange, open, openHotlink]
  );

  return (
    <StyledInput
      onMenuOption={onMenuOption}
      menuOptions={dropdownOptions}
      openMediaPicker={open}
      ref={forwardedRef}
      value={value === MULTIPLE_VALUE ? null : value}
      {...rest}
    />
  );
};

function MediaInput(
  {
    buttonInsertText = __('Choose an image', 'web-stories'),
    hotlinkTitle = __('Use external image', 'web-stories'),
    hotlinkInsertText = __('Use image', 'web-stories'),
    hotlinkInsertingText = __('Using image', 'web-stories'),
    menuOptions = [],
    onChange,
    onChangeErrorText = __(
      'Unable to use this file type, please select a valid image type.',
      'web-stories'
    ),
    title = __('Choose an image', 'web-stories'),
    type = 'image',
    value,
    cropParams,
    canUseProxy = false,
    ...rest
  },
  forwardedRef
) {
  const { MediaUpload } = useConfig();
  const {
    actions: { onSelect, openHotlink, onCloseHotlink },
    state: { allowedFileTypes, isOpen },
  } = useHotlink({ onChange, type, canUseProxy });

  const renderMediaIcon = useCallback(
    (open) => {
      // Options available for the media input menu.
      const availableMenuOptions = [
        { label: __('Upload a file', 'web-stories'), value: 'upload' },
        { label: __('Link to a file', 'web-stories'), value: 'hotlink' },
        { label: __('Remove', 'web-stories'), value: 'remove' },
        { label: __('Reset', 'web-stories'), value: 'reset' },
      ];

      // No menu for mixed value.
      // Match the options from props, if none are matched, menu is not displayed.
      const dropdownOptions =
        value === MULTIPLE_VALUE
          ? []
          : availableMenuOptions.filter(({ value: option }) =>
              menuOptions.includes(option)
            );
      return (
        <MediaInputField
          open={open}
          openHotlink={openHotlink}
          onChange={onChange}
          dropdownOptions={dropdownOptions}
          forwardedRef={forwardedRef}
          value={value}
          {...rest}
        />
      );
    },
    [value, openHotlink, onChange, forwardedRef, rest, menuOptions]
  );

  return (
    <>
      <HotlinkModal
        isOpen={isOpen}
        title={hotlinkTitle}
        onSelect={onSelect}
        onClose={onCloseHotlink}
        insertText={hotlinkInsertText}
        insertingText={hotlinkInsertingText}
        allowedFileTypes={allowedFileTypes}
        canUseProxy={canUseProxy}
        requiredImgDimensions={cropParams}
      />
      <MediaUpload
        title={title}
        buttonInsertText={buttonInsertText}
        onSelect={onChange}
        onSelectErrorMessage={onChangeErrorText}
        type={type}
        cropParams={cropParams}
        render={renderMediaIcon}
      />
    </>
  );
}

const MediaInputWithRef = forwardRef(MediaInput);

MediaInputField.propTypes = {
  open: PropTypes.func,
  openHotlink: PropTypes.func,
  dropdownOptions: PropTypes.array,
  onChange: PropTypes.func,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.elementType }),
  ]),
  value: PropTypes.string,
};

MediaInput.propTypes = {
  className: PropTypes.string,
  buttonInsertText: PropTypes.string,
  hotlinkTitle: PropTypes.string,
  hotlinkInsertText: PropTypes.string,
  hotlinkInsertingText: PropTypes.string,
  menuOptions: PropTypes.array,
  onChange: PropTypes.func.isRequired,
  onChangeErrorText: PropTypes.string,
  type: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  cropParams: PropTypes.object,
  canUseProxy: PropTypes.bool,
  title: PropTypes.string,
  value: PropTypes.string,
};

export default MediaInputWithRef;
