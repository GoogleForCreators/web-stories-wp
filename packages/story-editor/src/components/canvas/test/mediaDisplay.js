/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { render, act } from '@testing-library/react';
import { resourceList } from '@googleforcreators/media';
import { registerElementType } from '@googleforcreators/elements';
import { elementTypes } from '@googleforcreators/element-library';

/**
 * Internal dependencies
 */
import { TestDisplayElement } from './_utils';

/* eslint-disable testing-library/no-node-access, testing-library/no-container */

describe('MediaDisplay', () => {
  let imageElement;
  let videoElement;
  let storyContext;
  let refs;

  beforeAll(() => {
    elementTypes.forEach(registerElementType);
  });

  beforeEach(() => {
    imageElement = {
      id: 'foo',
      type: 'image',
      mimeType: 'image/png',
      x: 0,
      y: 0,
      width: 80,
      height: 100,
      flip: {
        vertical: false,
        horizontal: false,
      },
      rotationAngle: 0,
      scale: 100,
      focalX: 50,
      focalY: 50,
      isBackground: false,
      resource: {
        id: 123,
        type: 'image',
        mimeType: 'image/png',
        src: 'https://example.com/image1.png',
        width: 1000,
        height: 800,
        alt: '',
        sizes: {
          medium: {
            sourceUrl: 'https://example.com/image1-mid.png',
            width: 500,
            height: 400,
            mimeType: 'image/png',
          },
          full: {
            sourceUrl: 'https://example.com/image1.png',
            width: 1000,
            height: 800,
            mimeType: 'image/png',
          },
        },
      },
    };
    videoElement = {
      id: 'baz',
      type: 'video',
      mimeType: 'video/mp4',
      scale: 1,
      origRatio: 9 / 16,
      x: 0,
      y: 0,
      height: 1920,
      width: 1080,
      rotationAngle: 0,
      cale: 100,
      focalX: 50,
      focalY: 50,
      isBackground: false,
      loop: true,
      flip: {
        vertical: false,
        horizontal: false,
      },
      resource: {
        type: 'video',
        mimeType: 'video/mp4',
        id: 123,
        src: 'https://example.com/video.mp4',
        poster: 'https://example.com/poster.png',
        height: 1920,
        width: 1080,
        length: 99,
        alt: '',
      },
    };

    storyContext = {};
    refs = {};
  });

  it('should render img with srcset, when fullsize resource is loaded', () => {
    resourceList.set(imageElement.resource.id, { type: 'fullsize' });
    const { container } = render(
      <TestDisplayElement storyContext={storyContext} element={imageElement} />
    );

    const img = container.querySelector('img');

    expect(img.srcset).toBe(
      'https://example.com/image1.png 1000w,https://example.com/image1-mid.png 500w'
    );
    // Take optimized image loading into account, fullsize uses original image
    expect(img.src).toBe(imageElement.resource.src);
  });

  it('should render img with scale and focal point', () => {
    const { container } = render(
      <TestDisplayElement storyContext={storyContext} element={imageElement} />
    );

    const img = container.querySelector('img');
    // The offset is setup to fit 1000:800 image inside 80:100 box.
    expect(window.getComputedStyle(img)).toMatchObject({
      width: '125px',
      height: '100px',
      left: '-22.5px',
      top: '0px',
    });
  });

  it('should apply and reset transformations', () => {
    const { container } = render(
      <TestDisplayElement
        refs={refs}
        storyContext={storyContext}
        element={imageElement}
      />
    );

    const {
      actions: { pushTransform },
    } = refs.transformContext;

    const img = container.querySelector('img');

    // Start with empty style.
    expect(img.style).toMatchSnapshot('empty style before');

    // Resize to 100:80 (or 1:1 with the original size).
    act(() => {
      pushTransform(imageElement.id, { resize: [100, 80] });
    });
    expect(img).toHaveStyle({
      width: '100px',
      height: '80px',
      left: '0px',
      top: '0px',
    });

    // Reset.
    act(() => {
      pushTransform(imageElement.id, null);
    });
    expect(img.style).toMatchSnapshot('empty style after');
  });

  it('should render flipped background video', () => {
    const flipCases = [
      {
        flip: { vertical: true, horizontal: false },
        transform: 'scale3d(1, -1, 1)',
      },
      {
        flip: { vertical: false, horizontal: true },
        transform: 'scale3d(-1, 1, 1)',
      },
      {
        flip: { vertical: true, horizontal: true },
        transform: 'scale3d(-1, -1, 1)',
      },
    ];

    flipCases.forEach(({ flip, transform }) => {
      const flippedBackgroundVideo = {
        ...videoElement,
        isBackground: true,
        flip,
      };
      const { container } = render(
        <TestDisplayElement
          storyContext={{
            ...storyContext,
            page: {},
          }}
          element={flippedBackgroundVideo}
        />
      );

      const element = container.querySelector(
        '[data-element-id="baz"]'
      ).firstChild;
      expect(window.getComputedStyle(element)).toMatchObject({
        transform,
      });
    });
  });
});

/* eslint-enable testing-library/no-node-access, testing-library/no-container */
