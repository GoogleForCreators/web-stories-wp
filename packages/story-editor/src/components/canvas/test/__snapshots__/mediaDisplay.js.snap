// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaDisplay should apply and reset transformations: empty style after 1`] = `
CSSStyleDeclaration {
  "_importants": {},
  "_length": 0,
  "_onChange": [Function],
  "_values": {},
}
`;

exports[`MediaDisplay should apply and reset transformations: empty style before 1`] = `
CSSStyleDeclaration {
  "_importants": {},
  "_length": 0,
  "_onChange": [Function],
  "_values": {},
}
`;
